### PR TITLE
Use `indexmap` to avoid sorting `LocalDefId`s

### PR DESCRIPTION
**Pull request by pierwill**
See discussion in https://github.com/rust-lang/rust/pull/90408#discussion_r745935459.

Related to work on https://github.com/rust-lang/rust/issues/90317.

### CODE REVIEW

**Comment by cuviper** [comment]
FWIW, `indexmap` does have a feature support the normal `rayon` traits, but not the forked `rustc-rayon`.

**Comment by pierwill** [comment]
> FWIW, `indexmap` does have a feature to support the normal `rayon` traits, but not the forked `rustc-rayon`.

@cuviper Would we need to fork `indexmap` in order to wire it up to `rustc-rayon`? (Probably not what we want.) Or might there be a way to accomplish this on the rustc side?

Right now, I wouldn't say this is critical to fixing https://github.com/rust-lang/rust/issues/90317 https://github.com/rust-lang/rust/issues/90317. But it seems like something we'd want eventually.

**Comment by joshtriplett** [comment]
cc @nnethercote 

Long term, I think we want to resolve the rustc-rayon vs rayon fork, which would address the problem. (For instance, we could add an `unstable` feature flag to `rayon` to enable the features rustc needs.)

**Comment by cuviper** [comment]
> (For instance, we could add an `unstable` feature flag to `rayon` to enable the features rustc needs.)

Hmm... For a while we had `--cfg rayon_unstable` (**not** a cargo feature) that enabled support for the experimental `rayon-futures` crate, but we removed that in rayon-rs/rayon#716. I don't feel thrilled about heading back in that direction, but it probably would be better than letting this fork continue forever.

**Comment by wesleywiser** [comment]
@pierwill I see this is in "draft" mode. Do you want someone to review now or are you still working on making changes?

**Comment by pierwill** [comment]
@wesleywiser This is waiting on https://github.com/bluss/indexmap/pull/212. Feedback is always welcome, though!

**Comment by wesleywiser** [comment]
Got it, just wanted to make sure this wasn't blocked on me. 🙂

File: src/tools/tidy/src/extdeps.rs
Comment (pierwill) [inline src/tools/tidy/src/extdeps.rs:9]: Remove before approval/merge.

File: Cargo.toml
Comment (pierwill) [inline Cargo.toml:127]: Update to new version when released.

Comment (cuviper) [reply Cargo.toml:127]: I've now published indexmap 1.8.0.

**Comment by pierwill** [comment]
Running into a [licensing error with `crossbeam-queue`](https://github.com/rust-lang/rust/runs/4743863784?check_suite_focus=true#step:25:2187).

Could this be caused by the `indexmap` update, @cuviper?

**Comment by cuviper** [comment]
I think this is a bug in tidy.

In `src/tools/tidy/src/deps.rs` `EXCEPTIONS`, there is already an exception for `crossbeam-queue`. However, exceptions are not allowed for the runtime crates that make up the standard library.

With the `backtrace` feature, `std` depends on `object`. Then `object` has a conditional dependency on `indexmap`, but this is not enabled under `std`. However, if you look at the whole workspace metadata at once (as tidy does), then the compiler enables the `object` features that require `indexmap`, and now also the `indexmap` feature for `rustc-rayon` too.

So tidy sees `std` -> `object` -> `indexmap` -> `rustc-rayon` -> `rustc-rayon-core` -> `crossbeam-queue` in the metadata, and thus includes *all* of that in "runtime" crates, and then the exception isn't allowed. But if you look at `cargo tree --all-features -p std`, they aren't really included at all.

**Comment by pierwill** [comment]
Marking this as blocked on https://github.com/rust-lang/rust/issues/92677. @rustbot label +S-blocked

File: src/test/run-make/const_fn_mir/dump.mir
Comment (pierwill) [inline src/test/run-make/const_fn_mir/dump.mir:28]: Not sure if I was supposed to bless this test, or if it represents an error...

Comment (wesleywiser) [reply src/test/run-make/const_fn_mir/dump.mir:28]: This is fine!

**Review by wesleywiser — APPROVED** [review]

**Comment by pierwill** [comment]
> 💥 Test timed out

@wesleywiser Can we give this another try?

**Comment by wesleywiser** [comment]
Yeah seems unrelated. 

@bors retry

**Comment by pierwill** [comment]
@wesleywiser I just rebased this to get rid of a very silly commit (that said `bless test?` 😬 ). One more + should do it! Thank you— 🙏 😇

**Comment by pierwill** [comment]
@wesleywiser 🏁 Conflicts addressed. @rustbot ready

**Pull request merged**
